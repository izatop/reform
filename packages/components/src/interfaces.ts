import {AnchorHTMLAttributes, ButtonHTMLAttributes, DetailedHTMLProps} from "react";
import {Breakpoint} from "./enum";

export interface IDefaults {
    className?: string;
}

export interface IBreakpoint {
    breakpoint?: Breakpoint;
}

export type MakeProps<P = {}> = IDefaults & P;

export type MakeElementProps<P = {}, E = {}> = MakeProps<P> & {props?: E};

export type MakeBreakpointProps<P = {}> = MakeProps<P> & IBreakpoint;

export type MakeResponsive<T extends object, K extends keyof T> = T & {
    mobile?: Pick<T, K>;
    tablet?: Pick<T, K>;
    touch?: Pick<T, K>;
    desktop?: Pick<T, K>;
    widescreen?: Pick<T, K>;
    fullhd?: Pick<T, K>;
};

export type ITagButtonProps = DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement>;
export type ITagAnchorProps = DetailedHTMLProps<AnchorHTMLAttributes<HTMLAnchorElement>,
    HTMLAnchorElement>;
