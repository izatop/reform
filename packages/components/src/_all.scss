@charset "utf-8";

// font awesome
$fa-font-path: "~@fortawesome/fontawesome-free/webfonts";
@import "~@fortawesome/fontawesome-free/scss/fontawesome";
@import "~@fortawesome/fontawesome-free/scss/solid";

@import "~bulma/sass/utilities/all";
@import "~bulma/sass/base/all";
@import "~bulma/sass/form/all";
@import "~bulma/sass/elements/all";
@import "~bulma/sass/components/all";
@import "~bulma/sass/layout/all";
@import "~bulma/sass/grid/all";
