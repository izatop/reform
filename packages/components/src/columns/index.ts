export * from "./Column";
export * from "./Columns";
