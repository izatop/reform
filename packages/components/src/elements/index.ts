
export * from "./Button";
export * from "./Icon";
export * from "./Image";
export * from "./Notification";
export * from "./Progress";
export * from "./Table";
export * from "./Tag";
export * from "./Title";
export * from "./Box";
export * from "./Container";
export * from "./Content";
export * from "./Delete";
