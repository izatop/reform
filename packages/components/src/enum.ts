export enum Breakpoint {
    Mobile = "mobile",
    Tablet = "tablet",
    Touch = "touch",
    Desktop = "desktop",
    Widescreen = "widescreen",
    Fullhd = "fullhd",
}

export enum ContainerView {
    Fluid = "fluid",
}

export enum TextTransform {
    Capitalize = "capitalized",
    Lowercase = "lowercase",
    Uppercase = "uppercase",
    Italic = "italic",
}

export enum FontWeight {
    Light = "light",
    Normal = "normal",
    Semibold = "semibold",
    Bold = "bold",
}

export enum FontFamily {
    SansSerif = "sans-serif",
    Monospace = "monospace",
    Primary = "primary",
    Secondary = "secondary",
    Code = "code",
}

export enum TextSize {
    XSmall = 7,
    small = 6,
    Normal = 5,
    Medium = 4,
    Large = 3,
    XLarge = 2,
    XXLarge = 1,
}

export enum Size {
    Small = "small",
    Medium = "medium",
    Large = "large",
}

export enum Helper {
    Clearfix = "clearfix",
    PullLeft = "pulled-left",
    PullRight = "pulled-right",
    Marginless = "marginless",
    Paddingless = "paddingless",
    Overlay = "overlay",
    Clipped = "clipped",
    Radiusless = "radiusless",
    Shadowless = "shadowless",
    Unselectable = "unselectable",
    Invisible = "invisible",
    SROnly = "sr-only",
}

export enum Display {
    Block = "block",
    Flex = "flex",
    Inline = "inline",
    InlineBlock = "inline-block",
    InlineFlex = "inline-flex",
}

export enum Color {
    White = "white",
    Black = "black",
    Light = "light",
    Dark = "dark",
    Primary = "primary",
    Info = "info",
    Link = "link",
    Success = "success",
    Warning = "warning",
    Danger = "danger",
    BlackBis = "black-bis",
    BlackTer = "black-ter",
    GreyDarker = "grey-darker",
    GreyDark = "grey-dark",
    Grey = "grey",
    GreyLigth = "grey-light",
    GreyLighter = "grey-lighter",
    WhiteTer = "white-ter",
    WhiteBis = "white-bis",
}

export enum TextAlign {
    Center = "centered",
    Justify = "justified",
    Left = "left",
    Right = "right",
}

export type Align = "centered" | "right";
