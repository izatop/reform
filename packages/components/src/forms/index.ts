export * from "./Form/Control";
export * from "./Form/Field";
export * from "./Form/FieldHelp";
export * from "./Form/Label";
export * from "./Form/Element";
