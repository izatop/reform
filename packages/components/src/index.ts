export * from "./columns";
export * from "./components";
export * from "./elements";
export * from "./layout";
export * from "./enum";
export * from "./interfaces";
