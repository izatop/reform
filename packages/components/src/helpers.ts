import {Breakpoint} from "./enum";

/**
 * @private
 */
export interface ICalcResolver {
    [key: string]: (value: any) => string | string[] | undefined;
}

/**
 * @private
 */
export interface ICalcOptions {
    name?: string;
    responsive?: boolean;
    are?: Array<ICalcResolver | string>;
    has?: Array<ICalcResolver | string>;
    is?: Array<ICalcResolver | string>;
    props?: string[];
}

/**
 * @private
 */
export interface IProps {
    [key: string]: any;
}

const mergePropsAllow = ["className"];

export class Helpers {
    public static getClasses(types: {[key: string]: Array<ICalcResolver | string>}, props: IProps, suffix = "") {
        const classes = [];
        for (const [prefix, modifiers] of Object.entries(types)) {
            for (const item of modifiers) {
                if (typeof item === "string" && Reflect.has(props, item)) {
                    const value = Reflect.get(props, item);

                    if (typeof value === "string" && value) {
                        classes.push(`${prefix}-${value}${suffix}`);
                    } else if (typeof value === "boolean" && value) {
                        classes.push(`${prefix}-${item.toLowerCase()}${suffix}`);
                    }

                    continue;
                }

                if (typeof item === "object") {
                    for (const [key, resolve] of Object.entries(item)) {
                        if (Reflect.has(props, key)) {
                            const result = resolve(Reflect.get(props, key));
                            classes.push(
                                ...(Array.isArray(result) ? result : [result])
                                    .filter((value) => !!value)
                                    .map((name) => `${prefix}-${name}${suffix}`),
                            );
                        }
                    }
                }
            }
        }

        return classes;
    }

    public static calcClasses(props: IProps, options: ICalcOptions = {}) {
        const classes = [];

        const types = {
            are: options.are || [],
            has: options.has || [],
            is: options.is || [],
        };

        classes.push(...Helpers.getClasses(types, props));

        if (options.name) {
            classes.push(options.name);
        }

        if (options.responsive) {
            for (const device of Object.keys(Breakpoint)) {
                classes.push(...Helpers.getClasses(
                    types,
                    Reflect.get(props, device) || {},
                    `-${device}`,
                ));
            }
        }

        if (Reflect.has(props, "className")) {
            classes.push(Reflect.get(props, "className"));
        }

        return classes
            .filter((className) => !!className)
            .join(" ");
    }

    public static calcProps(props: IProps, options: ICalcOptions) {
        return {
            className: Helpers.calcClasses(props, options),
            ...(props.props || {}),
        };
    }

    public static mergeProps(...props: IProps[]) {
        const newProps: any = {};
        for (const part of props) {
            for (const [key, value] of Object.entries(part)) {
                newProps[key] = value;
            }
        }

        for (const key of mergePropsAllow) {
            newProps[key] = "";
            for (const part of props) {
                if (key in part) {
                    newProps[key] = newProps[key].concat(" ", part[key]);
                }
            }
        }

        return newProps;
    }
}
