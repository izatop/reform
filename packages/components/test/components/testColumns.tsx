import * as React from "react";
import * as renderer from "react-test-renderer";
import {Columns} from "../../src";

test("Columns", () => {
    const columns1 = renderer.create((
        <Columns is-gap={false} is-breakpoint={"fullhd"}/>
    ));

    const columns2 = renderer.create((
        <Columns is-gap={8} x-mobile={{"is-gap": 1}}/>
    ));

    expect(columns2.toJSON()).toMatchSnapshot();
    expect(columns1.toJSON()).toMatchSnapshot();
});
