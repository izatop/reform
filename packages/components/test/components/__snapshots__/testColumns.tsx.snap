// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns 1`] = `
<div
  className="columns is-8 is-1-mobile"
/>
`;

exports[`Columns 2`] = `
<div
  className="columns is-gapless is-fullhd"
/>
`;
